addons/product/models/product_template.py

```diff
@@ -489,6 +484,20 @@ def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_ui
             '', args=[('id', 'in', list(searched_ids))],
             operator='ilike', limit=limit, name_get_uid=name_get_uid)
 
+    def _get_ids_of_tmpl_without_variant(self):
```

Comment:
We should not have a method that returns "all" of something, this looks like a scaling disaster in terms of performance.

At best it should just return a subquery to inject into another query, but then it should just be part of the original query, not a separate method.

Reply:
in 14.0 it will be a subquery (and no direct SQL). in 13.0 I'm not sure how this could be done differently, so this optimized code (only one query with join returning only the record without variants or with only archived variant that should be rather rare) is to avoid the scaling disaster.

The current solution could be combined with the name search:

```diff
- def _get_ids_of_tmpl_without_variant(self):
+ def _get_ids_of_tmpl_without_variant(self, domain=None, limit, order, ...):
-    query = self.env['product.template']._where_calc([])
+    query = self.env['product.template']._where_calc(args or [])
+ # also implement limit, sort, ... 
```
but then we need to reimplement _name_search a little:

````py   
name_search_domain = [(self._rec_name, operator, name)] if self._rec_name and (name, operator) != ('', 'ilike') else []
result = _get_ids_of_tmpl_without_variant(domain or [] + name_search_domain, limit, order, name_get_uid);
```
so this would complicate the current solution (that we did not want and wanted to replace with simpler query that is doing something similar than what is currently done) for something that is only in 13.0 and only possibly slightly improving case of thousands of products without any variants which should be rare.